### PR TITLE
Update the data structure of problem instances

### LINKED ISSUE #119: Update data structure of problem instance
### Feature Description
**Please describe the feature you would like to see added:**

Based on user expectation, the structure of the instance file for the BoxQP problem needs to be updated:
- The first row of the instance file (`instance_info`) has additional fields: `best_sol` in 3rd col,  `sol_time_bfgs` in 6th col, `num_frac_values` in 8th col
- Compared the old data structure we have an additional row at the end of the new instance file.

### Rationale
**Why is this feature important? What problem does it solve?**

Aforementioned additional fields are required for benchmarking purposes. That is, they will be used as a baseline for comparing the outcome of the solvers

### PR DESCRIPTION
**Pull request by MehmetC-1QBit**
### Description

> Describe the purpose of this PR along with any background information and the impacts of the proposed change. For the benefit of the community, please do not assume prior context.

The data structure of the instance files need to be updated to address end user expectations.
The change is not vital for the solvers but required for further investigation during the data analysis.

Here is the summary of the change:
1. The old instance file in `examples/test_instances/` is replaced with new instance
2. The old instance file in `examples/tuning_instances/` is replaced with new instance 
3.  Based on #119, 
    - `ProblemInstance` is updated (including docstring) in `ccvm_simulators/problem_classes/boxqp/problem_instance.py`
    - `Solution` is updated (including docstring) in `ccvm_simulators/solution.py`
4. `CHANGELOG.md` is updated
5. `README.md` is fixed to reflect the change from previous PR
6. `ccvm_simulators/problem_classes/README.md` is updated
7. Also `examples/README.md` is updated similar to item 6 (Reviewer's attention is required the instance info is redundant with item 6)
8. The modules `dl_solver.py`, `mf_solver.py`, and `langevin_solver.py` in `ccvm_simulators/solvers/` are updated

### References

Closes #119 

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `main`

### CODE REVIEW

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py [hunk 1]
```diff
@@ -143,10 +157,18 @@ def load_instance(
...
                     rval_v[idx] = -torch.Tensor([float(line_data_v[idx])])
                 # Read in the q_matrix matrix
-                for idx, line in enumerate(lines[2:]):
+                for idx, line in enumerate(lines[2 : problem_size + 2]):
                     line_data = line.split("\n")[0].split(file_delimiter)
                     for j, value in enumerate(line_data[:problem_size]):
                         rval_q[idx, j] = -torch.Tensor([float(value)])
+
+                # Read the last line as an additional information
+                last_raw_data_line = lines[-1].split("\n")[0].split(file_delimiter)
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/problem_classes/boxqp/problem_instance.py:166]: We'll need to be more explicit, since this line is optional. Otherwise, if the user leaves out the final line then optional_data will become populated with the last line of q_matrix.
```suggestion
                last_raw_data_line = lines[problem_size + 3].split("\n")[0].split(file_delimiter)
```

**Review by KatieO4t1qbit — CHANGES_REQUESTED** [review]
Thanks for this change, Mehmet!
Let's ensure that the tests pass as part of this change by running `pytest ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py` (assumes you are in the top-level directory)

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py [hunk 2]
```diff
@@ -143,10 +157,18 @@ def load_instance(
...
+                optional_info = []
+                for v in last_raw_data_line:
+                    if not v == "":
+                        optional_info.append(float(v))
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/problem_classes/boxqp/problem_instance.py:170]: Give this a descriptive name such as `solution_vector` instead of `optional info`.

File: ccvm_simulators/solution.py [hunk 3]
```diff
@@ -18,6 +18,9 @@ class Solution:
         solve_time (float): Time to solve the problem.
         pp_time (float):  Time to post-process the problem.
         optimal_value (float): The optimal objective value for the given problem instance.
+        best_value (float): The best objective value for the given problem instance.
+        num_frac_values (int): Number of fractional values in the solutions of given problem instance.
+        optional_info (list): Optional information in a given problem instance.
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/solution.py:23]: Similar change here. Describe as the solution to the problem instance instead of optional information.

File: ccvm_simulators/solvers/dl_solver.py [hunk 4]
```diff
@@ -433,6 +435,9 @@ def _solve(
             solve_time=solve_time,
             pp_time=pp_time,
             optimal_value=instance.optimal_sol,
+            best_value=instance.best_sol,
+            num_frac_values=instance.num_frac_values,
+            optional_info=instance.optional_info,
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/solvers/dl_solver.py:440]: Apply the previous comments here as well.

File: ccvm_simulators/solvers/langevin_solver.py [hunk 5]
```diff
@@ -380,6 +383,9 @@ def _solve(
             solve_time=solve_time,
             pp_time=pp_time,
             optimal_value=instance.optimal_sol,
+            best_value=instance.best_sol,
+            num_frac_values=instance.num_frac_values,
+            optional_info=instance.optional_info,
```
Comment (FarhadK-1QBit) [inline ccvm_simulators/solvers/langevin_solver.py:388]: And in this file.

File: examples/README.md [hunk 6]
```diff
@@ -20,10 +20,18 @@ The following folders contain our example problem instances:
...
+- solution time for BFGS to solve it
+- the seed number used for generating the instance in `torch`.
+- number of fractional values in the solution
 
-The second line of the instance file contains the elements of the vector $V$, which defines the instance. The rest of the lines in the file contain the matrix $Q$.
+The second line contains the elements of the vector `V`, which describes the instance.
+
+The rest of the lines before the last line hold the elements of the `Q` matrix.
+
+The final line contains list of additional (optional) information in the problem instance.
```
Comment (FarhadK-1QBit) [inline examples/README.md:37]: Say "the final line contains the vector of the solution to the problem instance".

**Review by FarhadK-1QBit — CHANGES_REQUESTED** [review]
Thanks for implementing this change, Mehmet! 
I only have some minor comments and changes regarding the last line of the problem instance.

File: CHANGELOG.md [hunk 7]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Updated internal data handling that have improved performance and reduced memory usage.
+- Updated data structure of the problem instance
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply CHANGELOG.md]: I added the name of the additional parameters in the changelog.md.

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py [hunk 8]
```diff
@@ -126,11 +135,16 @@ def load_instance(
...
+                if instance_info[3].lower() == "true":
                     optimality = True
                 else:
                     optimality = False
-                sol_time_gb = float(instance_info[3])
+                sol_time_gb = float(instance_info[4])
+                sol_time_bfgs = float(instance_info[5])
+                num_frac_values = int(
+                    instance_info[7]
+                )  # seed=int(instance_info[6]) # discarded
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/boxqp/problem_instance.py:147]: Well! Based on our discussion I just disregarded it because we have already `seed` info from the instance file name.

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py [hunk 9]
```diff
@@ -44,19 +44,28 @@ def __init__(
...
+            optimal_sol (float): the optimal solution via Gurobi to the problem. Defaults to None.
+            best_sol (float): the best solution via BFGS. Defaults to None
             optimality (bool): indicates whether the solution is
                 optimal (True or False). Defaults to None.
+            sol_time_gb (float): the time for Gurobi to solve it
+            sol_time_bfgs (float): the time for BFGS to solve it
+            num_frac_values (int): number of fractional values in the solution
             q_matrix (torch.tensor): Q matrix of the QP problem. Defaults to None.
             v_vector (torch.tensor): V vector of the QP problem. Defaults to None.
+            optional_info (list): additional information in the problem instance
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/boxqp/problem_instance.py]: - @KatieO4t1qbit  I guess @FarhadK-1QBit 's answer is good enough to move forward:)
- Updated docstring of the field `optional_info` based on @FarhadK-1QBit 's comment.

**Comment by MehmetC-1QBit** [comment]
> Thanks for this change, Mehmet! Let's ensure that the relevant tests pass as part of this change by running `pytest ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py` (assumes you are in the top-level directory)

I updated test modules and replaced old instance files.

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py (hunk 1 above)
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/boxqp/problem_instance.py:166]: Any suggestion for this?

File: ccvm_simulators/problem_classes/README.md [hunk 10]
```diff
@@ -31,12 +31,18 @@ independet of the particular random instances generated.
...
+- solution time for BFGS to solve it
 - the seed number used for generating the instance in `torch`.
+- number of fractional values in the solution
 
 The second line contains the elements of the vector `V`, which describes the instance.
 
-The rest of the lines hold the elements of the `Q` matrix.
+The rest of the lines before the last line hold the elements of the `Q` matrix.
+
+The final line contains list of additional (optional) information in the problem instance.
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/README.md:45]: Done

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py (hunk 2 above)
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/boxqp/problem_instance.py]: done

File: ccvm_simulators/solution.py (hunk 3 above)
Comment (MehmetC-1QBit) [reply ccvm_simulators/solution.py]: done

File: ccvm_simulators/solvers/langevin_solver.py (hunk 5 above)
Comment (MehmetC-1QBit) [reply ccvm_simulators/solvers/langevin_solver.py]: done

File: problem_variables.pt
Comment (KatieO4t1qbit) [inline problem_variables.pt:1]: What is this file?

File: CHANGELOG.md [hunk 11]
```diff
@@ -18,6 +18,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Updated internal data handling that have improved performance and reduced memory usage.
+- Updated data structure of the problem instance file:
+	-- To extract all the values in the first line of the instance, additional fields i.e. (`best_sol`, `sol_time_bfgs`, `num_frac_values`) are added.
+	--  The field`solution_vector` is added to extract vector of the solution to the problem instance at the end of the new instance file compared the old data structure
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply CHANGELOG.md]: To answer your questions, let's keep in mind that we are extending the extraction of input data from a new problem instance. These fields are added because end users will have an option for further investigation.
Anyway, thank you for the suggestion: I modified your answer and rewrite both terms

File: ccvm_simulators/solution.py [hunk 12]
```diff
@@ -18,6 +18,9 @@ class Solution:
         solve_time (float): Time to solve the problem.
         pp_time (float):  Time to post-process the problem.
         optimal_value (float): The optimal objective value for the given problem instance.
+        best_value (float): The best objective value for the given problem instance.
+        num_frac_values (int): Number of fractional values in the solutions of given problem instance.
+        solution_vector (list): Optional information in a given problem instance.
```
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solution.py]: Thanks for the modification. I put in the doctstring except `Optional.`

File: problem_variables.pt
Comment (MehmetC-1QBit) [reply problem_variables.pt:1]: I remember such file name was in one of test cases. It was unintentionally created during debugging session. 
Now deleted.

File: ccvm_simulators/solution.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply ccvm_simulators/solution.py]: Thanks for the suggestion. I updated the line 23 but this time I excluded "Optional." Because the new input file has it.

File: CHANGELOG.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (MehmetC-1QBit) [reply CHANGELOG.md]: Thanks for the catch. I updated that line based on your suggestion. Please check out the commit

File: ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py [hunk 13]
```diff
@@ -97,19 +97,48 @@ def test_load_instance_cpu(self):
...
+        self.problem_size = problem_size
+        self.optimal_sol = optimal_sol
+        self.best_sol = best_sol
+        self.optimality = optimality
+        self.sol_time_gb = sol_time_gb
+        self.sol_time_bfgs = sol_time_bfgs
+        self.num_frac_values = num_frac_values
+        self.q_matrix = rval_q
+        self.v_vector = rval_v
+        self.optional_info = optional_info
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py:113]: I think this should be solution_vector now?

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py (hunk 1 above)
Comment (KatieO4t1qbit) [reply ccvm_simulators/problem_classes/boxqp/problem_instance.py:166]: I think you may have forgotten to push this change?

**Review by KatieO4t1qbit — APPROVED** [review]
Thanks Mehmet! Looking good 🔥 
I left a couple comments, please address them before merging. Thanks!

File: ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py (hunk 13 above)
Comment (MehmetC-1QBit) [reply ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py]: Thanks Katie, I removed the lines between 101-113 which was actually a commented segment that I copied from problem_instance.py but forgot to delete. Now it is done.

File: ccvm_simulators/problem_classes/boxqp/problem_instance.py [hunk 14]
```diff
@@ -143,10 +157,18 @@ def load_instance(
...
                     for j, value in enumerate(line_data[:problem_size]):
                         rval_q[idx, j] = -torch.Tensor([float(value)])
+
+                # Read the last line as an additional information
+                last_raw_data_line = lines[problem_size + 2].split("\n")[0].split(file_delimiter)
+                solution_vector = []
+                for v in last_raw_data_line:
+                    if not v == "":
+                        solution_vector.append(float(v))
+
```
Comment (KatieO4t1qbit) [inline ccvm_simulators/problem_classes/boxqp/problem_instance.py:171]: Is the solution vector is required before parsing? If it is optional we will need to wrap this in a try block to allow the problem to be read without error in the case that the solution vector is not given:
```suggestion
                 # Read the last line as an additional information
                solution_vector = []
                try:
                    last_raw_data_line = lines[problem_size + 2].split("\n")[0].split(file_delimiter)
                    for v in last_raw_data_line:
                        if not v == "":
                            solution_vector.append(float(v))
                except IndexError:
                    # solution_vector was not supplied
                    pass
```

**Review by KatieO4t1qbit — APPROVED** [review]
Looks great! Thank you, Mehmet 🙌

**Review by FarhadK-1QBit — APPROVED** [review]
This Look great, Mehmet! Thank you.
